pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -617,25 +640,35 @@ func (r *ReconcileMachinePool) syncMachineSets(
 				if rMS.Spec.Template.Spec.Taints == nil {
 					rMS.Spec.Template.Spec.Taints = []corev1.Taint{}
 				}
-				// Make a temporary map of generated MachineSet's taints for easy lookup
-				gTaintsByKey := make(map[string]*corev1.Taint)
-				for gIndex, gTaint := range ms.Spec.Template.Spec.Taints {
-					gTaintsByKey[gTaint.Key] = &ms.Spec.Template.Spec.Taints[gIndex]
+				// First, build a list of remote MachineSet taints that do not have taints needing to be deleted
+				taints := make([]corev1.Taint, 0)
+				for _, rTaint := range rMS.Spec.Template.Spec.Taints {
+					if taintsToDelete.Has(identifierForTaint(&rTaint)) {
+						// This entry would be deleted from the final list
+						objectModified = true
+						continue
+					}
+					taints = append(taints, rTaint)
 				}
-				// Go through the remote MachineSet's taints, replacing overlaps with the generated MachineSet
-				for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
-					if gTaint, foundTaint := gTaintsByKey[rTaint.Key]; foundTaint {
-						if gTaint.Value != rTaint.Value || gTaint.Effect != rTaint.Effect {
-							rMS.Spec.Template.Spec.Taints[rIndex] = *gTaint
-							objectModified = true
+				rMS.Spec.Template.Spec.Taints = taints
+
+				for _, gTaint := range ms.Spec.Template.Spec.Taints {
+					// Taints are provided as a list, so duplicate entries are possible.
+					// Use a flag to indicate if a taint was found, append if not.
+					foundTaint := false
+					for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
+						if reflect.DeepEqual(identifierForTaint(&gTaint), identifierForTaint(&rTaint)) {
+							foundTaint = true
+							if gTaint.Value != rTaint.Value {
+								rMS.Spec.Template.Spec.Taints[rIndex].Value = gTaint.Value
```

Comment:
Another thing that's happening here: even for "owned" taints, we're preserving the order as it existed previously on the spoke. So the next bug report will be: "I can't change the order of taints by editing the MachinePool."

(I'll note that my attempt was also patently wrong on this count, as I was sorting.)

This only matters if taint order matters in the first place. I'm not sure it does. Research needed.

Reply:
Taint order doesn't matter as far as I can see, for the corresponding node should tolerate all taints from the corresponding Machine (though I'm still very green wrt taints and tolerations). That being said, I have removed the sorting of labels and taints - so the user is aware about which taints are collapsed.